File: Multicore/articulo/inverse.js
'use strict';

var MAXN = 150;
var numWorkers = 4;
var startTime;

var original = [];
original.length = MAXN;

var padWorkers = [];
padWorkers.length = numWorkers;

var gaussWorkers = [];
gaussWorkers.length = numWorkers;

function gaussJordan(A, i) {
  if (i === A.length) {
    var endTime = performance.now();
    if (numWorkers > 1) {
      $("#parallel").html((endTime - startTime) / 1000);
      //printMatrix(A);

      numWorkers = 1;
      padMatrix(original);
    } else {
      $("#sequential").html((endTime - startTime) / 1000);
      //printMatrix(A);
    }
    return;
  }

  var irow = A.shift();
  var val = irow[i];

  var workersFinished = 0;
  var start = 0;
  var chunkSize = MAXN * 2 / numWorkers;
  var end;

  for (var j = 0; j < numWorkers; j++) {
    end = j + 1 === numWorkers ? MAXN - 1 : Math.ceil(start + chunkSize);
    gaussWorkers[j].postMessage({start: start, end: end, irow: irow, val: val});
    start = end;

    gaussWorkers[j].onmessage = function (event) {
      workersFinished++;
      for (var k = event.data.start; k < event.data.end; k++) {
        irow[k] = event.data.irow[k];
      }

      if (workersFinished === numWorkers) {
        workersFinished = 0;
        start = 0;
        chunkSize = (MAXN - 1) / numWorkers;

        for (var k = 0; k < numWorkers; k++) {
          end = k + 1 === numWorkers ? MAXN - 1 : Math.ceil(start + chunkSize);
          gaussWorkers[k].postMessage({start: start, end: end, irow: irow, i: i, A: A, size: MAXN});
          start = end;

          gaussWorkers[k].onmessage = function (event) {
            workersFinished++;
            for (var l = event.data.start; l < event.data.end; l++) {
              A[l] = event.data.jrow[l];
            }
            if (workersFinished === numWorkers) {
              A.push(irow);
              gaussJordan(A, i + 1);
            }
          };
        }
      }
    }
  };
}

function padMatrix(matrix) {
  startTime = performance.now();
  var workersFinished = 0;
  var start = 0;
  var chunkSize = MAXN / numWorkers;
  var end;

  for (var i = 0; i < numWorkers; i++) {
    end = i + 1 === numWorkers ? MAXN - 1 : Math.ceil(start + chunkSize);
    padWorkers[i].postMessage({start: start, end: end, matrix: matrix, size: MAXN});
    start = end;

    padWorkers[i].onmessage = function (event) {
      workersFinished++;
      for (var j = event.data.start; j < event.data.end; j++) {
        matrix[j] = event.data.padded[j];
      }
      if (workersFinished === numWorkers) {
        gaussJordan(matrix, 0);
      }
    };
  }
}

function printMatrix(matrix) {
  var res = "";
  for (var i = 0; i < MAXN; i++) {
    for (var j = MAXN; j < MAXN * 2; j++) {
      res += matrix[i][j] + ", ";
    }
    res += "\n";
  }
  console.log(res);
}

$(function () {
  /*
  original = [
    [1.0, 1.0, 2.0, 1.0],
    [1.0, 2.0, 1.0, 2.0],
    [1.0, 1.0, 1.0, 3.0],
    [2.0, 3.0, 1.0, 3.0]];
  MAXN = 4;
  //*/

  //*
     var randomNum;
     for (var i = 0; i < MAXN; i++) {
     original[i] = [];
     original[i].length = MAXN;
     for (var j = 0; j < MAXN; j++) {
     randomNum = 1.0 + (Math.random() * MAXN);
     original[i][j] = randomNum;
     }
     }
  //*/

  for (var i = 0; i < numWorkers; i++) {
    padWorkers[i] = new Worker('pad_worker.js');
    gaussWorkers[i] = new Worker('gauss_worker.js');
  }

  padMatrix(original.slice());
});

File: Multicore/articulo/Inverse.java
/*
 * Authors:
 *  A01370815 Diego Galindez Barreda
 *  A01165988 Eduardo Azuri Gaytan Martinez 
 */
import java.util.ArrayList;
import java.util.Arrays;

public class Inverse {
    public static int MAXN = 1000;

    // Sequential code
    public static double[][] invert(double matrix[][]) {
        for (int i = 0; i < MAXN; i++) {
            matrix[i][MAXN + i] = 1;
        }

        gaussJordan(matrix, 0);

        return matrix;
    }

    public static double[][] gaussJordan(double matrix[][], int i) {
        if (i == matrix.length) {
            return matrix;
        }

        double[] irow = new double[MAXN * 2];
        irow = Arrays.copyOf(matrix[0], MAXN * 2);
        double val = irow[i];
        for (int j = 0; j < MAXN * 2; j++) {
            irow[j] /= val;
        }

        for (int j = 1; j < MAXN; j++) {
            double[] jrow = matrix[j];
            double scale = matrix[j][i];
            for (int k = 0; k < MAXN * 2; k++) {
                jrow[k] -= (irow[k] * scale);
            }
        }
        for (int j = 0; j < MAXN - 1; j++) {
            matrix[j] = matrix[j + 1];
        }
        matrix[MAXN - 1] = new double[MAXN * 2];
        matrix[MAXN - 1] = Arrays.copyOf(irow, MAXN * 2);

        return gaussJordan(matrix, i + 1);
    }

    // Parallel code
    public static double[][] parallelInvert(double matrix[][]) {
        for (int i = 0; i < MAXN; i++) {
            matrix[i][MAXN + i] = 1;
        }

        parallelGaussJordan(matrix, 0);

        return matrix;
    }

    public static double[][] parallelGaussJordan(double matrix[][], int i) {
        if (i == matrix.length) {
            return matrix;
        }

        double[] irow = new double[MAXN * 2];
        irow = Arrays.copyOf(matrix[0], MAXN * 2);
        double val = irow[i];
        for (int j = 0; j < MAXN * 2; j++) {
            irow[j] /= val;
        }

        for (int j = 1; j < MAXN; j++) {
            double[] jrow = matrix[j];
            double scale = matrix[j][i];
            for (int k = 0; k < MAXN * 2; k++) {
                jrow[k] -= (irow[k] * scale);
            }
        }
        for (int j = 0; j < MAXN - 1; j++) {
            matrix[j] = matrix[j + 1];
        }
        matrix[MAXN - 1] = new double[MAXN * 2];
        matrix[MAXN - 1] = Arrays.copyOf(irow, MAXN * 2);

        return parallelGaussJordan(matrix, i + 1);
    }

    public static void printMatrix(double[][] matrix) {
        for (int i = 0; i < matrix.length; i++) {
            for (int j = MAXN; j < matrix[0].length; j++) {
                System.out.print(matrix[i][j] + ", ");
            }
            System.out.println();
        }
        System.out.println();
    }

    public static void benchmark(String title, double[][] matrix, boolean sequential) {
        double[][] inverse;

        System.out.println(title);
        //printMatrix(matrix);

        long start = System.currentTimeMillis();
        if (sequential) {
            invert(matrix);
        } else {
            parallelInvert(matrix);
        }
        long end = System.currentTimeMillis();
        double time = (end - start) / 1000.0;

        //printMatrix(matrix);
        System.out.printf("Time: %.2f%n", time);
    }

    public static void main(String... args) {
        //*
           double randomNum;
           double[][] matrix = new double[MAXN][MAXN * 2];
           for (int i = 0; i < MAXN; i++) {
               for (int j = 0; j < MAXN; j++) {
                   randomNum = 1.0 + (Math.random() * MAXN);
                   matrix[i][j] = randomNum;
               }
           }
           //*/
           /*
        double[][] matrix = {
            {1, 1, 2, 1, 0, 0, 0, 0},
            {1, 2, 1, 2, 0, 0, 0, 0},
            {1, 1, 1, 3, 0, 0, 0, 0},
            {2, 3, 1, 3, 0, 0, 0, 0}
        };
           */

        benchmark("Sequential", Arrays.copyOf(matrix, MAXN), true);
        benchmark("Concurrent", matrix, false);
    }
}
